### ISSUE TITLE
Forcing max concurrency from location block crashes passenger core

### ISSUE THREAD

**Issue by felixbuenemann** [phusion/passenger#1720]
I'm trying to run a rack app with the message_bus middleware in passenger nginx 5.0.23, so that the message_bus long polling endpoint has unlimited concurrency, while all other requests use default concurrency.

However this crashes the passenger core, as soon as a second concurrent request hits the unlimited concurrency location block.

The crash does not occur, if I set `PhusionPassenger.advertised_concurrency_level = 0` inside the config.ru, which overrides the setting from the nginx config.

Nginx config:
```
server {
        listen 80 default_server;
        listen [::]:80 default_server;

        root /home/felix/hijack_test/public;

        server_name _;

        passenger_ruby /home/felix/.rvm/wrappers/ruby-2.2.4/ruby;
        passenger_enabled on;
        passenger_app_group_name app;

        ## location ~ ^/message-bus/(.*/)?poll$ {
        location /message-bus {
                passenger_app_group_name message_bus;
                passenger_force_max_concurrent_requests_per_process 0;
        }
}
```

config.ru:
```ruby
require 'message_bus'

if defined?(PhusionPassenger)
    # PhusionPassenger.advertised_concurrency_level = 10000
    PhusionPassenger.on_event(:starting_worker_process) do |forked|
        if forked
            # We're in smart spawning mode.
            MessageBus.after_fork
        else
            # We're in conservative spawning mode. We don't need to do anything.
        end
    end
end

MessageBus.redis_config = { url: "redis://127.0.0.1:6379/7" }

MessageBus.long_polling_interval = 1000 * 25
MessageBus.max_active_clients = 10000

use MessageBus::Rack::Middleware
run lambda { |env| [200, {"Content-Type" => "text/html"}, [Time.now.to_s]]  }
```

Gemfile:
```ruby
source "https://rubygems.org"
gem "message_bus", github: 'SamSaffron/message_bus'
```

Testing (foo and bar are random client ids):
```
$ curl -iX POST localhost/message-bus/foo/poll
HTTP/1.1 200 OK
Content-Type: text/plain; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Status: 200 OK
Cache-Control: must-revalidate, private, max-age=0
Pragma: no-cache
X-Content-Type-Options: nosniff
Expires: 0
Date: Sat, 09 Jan 2016 22:22:42 GMT
X-Powered-By: Phusion Passenger 5.0.23
Server: nginx/1.8.0 + Phusion Passenger 5.0.23

$ curl -iX POST localhost/message-bus/bar/poll
HTTP/1.1 502 Bad Gateway
Server: nginx/1.8.0
Date: Sat, 09 Jan 2016 22:22:46 GMT
Content-Type: text/html
Content-Length: 172
Connection: keep-alive

<html>
<head><title>502 Bad Gateway</title></head>
<body bgcolor="white">
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.8.0</center>
</body>
</html>
```

Full crashlog: https://gist.github.com/2bc75c3ec63096c34247

Btw. gdb is installed in /usr/bin/gdb, but not detected by the crash watcher…

System: Ubuntu 14.04 LTS Server x86_64 and official passenger/nginx 5.0.23 ubuntu packages.

See SamSaffron/message_bus#61 for additional discussion.

**Comment by OnixGH** [comment]
Thanks for reporting. I think this is probably the same one I [encountered last friday](https://groups.google.com/forum/#!topic/phusion-passenger/GmchmCxGzi8). It looks like the option is not completely implemented: the nginx option sets concurrency at the process level, but not at the socket level, which results in the process not finding a socket to relay to -- whereas if the option is set at the application side, it is applied to the socket, and then upwards to the process. Should be fixable.

**Comment by OnixGH** [comment]
@felixbuenemann the branch GH-1720 fixes the crash for me (and the workers appear to go about their business normally), can you see if that's the case for you as well?

**Comment by felixbuenemann** [comment]
Sure, I should be able to test it today or tomorrow.

**Comment by felixbuenemann** [comment]
I can confirm that the bug is fixed in the GH-1720 branch.

**Comment by jandudulski** [comment]
Hi,

when can you release the fix?

**Comment by OnixGH** [comment]
@jandudulski we scheduled the fix for the next release (5.0.24). Will report back when we have a final date set for that.
